contrib/gnu/ed/ed.1
.TH ED 1 "10 November 1994"
.SH NAME
ed, red \- text editor
.SH SYNOPSIS
ed [-] [-Gs] [-p \fIstring\fR] [\fIfile\fR]
.LP
red [-] [-Gs] [-p \fIstring\fR] [\fIfile\fR]
.SH DESCRIPTION
.B ed
is a line-oriented text editor.
It is used to create, display, modify and otherwise manipulate text
files.
.B red
is a restricted
.BR ed :
it can only edit files in the current
directory and cannot execute shell commands.

If invoked with a
.I file
argument, then a copy of
.I file
is read into the editor's buffer.
Changes are made to this copy and not directly to
.I file
itself.
Upon quitting
.BR ed ,
any changes not explicitly saved  with a
.I `w'
command are lost.

Editing is done in two distinct modes:
.I command
and
.IR input .
When first invoked,
.B ed
is in command mode.
In this mode commands are read from the standard input and
executed to manipulate the contents of the editor buffer.
A typical command might look like:
.sp
.RS
,s/\fIold\fR/\fInew\fR/g
.RE
.sp
which replaces all occurences of the string
.I old
with
.IR new .

When an input command, such as
.I `a'
(append),
.I `i'
(insert) or
.I `c'
(change), is given,
.B ed
enters input mode.  This is the primary means
of adding text to a file.
In this mode, no commands are available;
instead, the standard input is written
directly to the editor buffer.  Lines consist of text up to and
including a
.IR newline
character.
Input mode is terminated by
entering a single period  (\fI.\fR) on a line.

All
.B ed
commands operate on whole lines or ranges of lines; e.g.,
the
.I `d'
command deletes lines; the
.I `m'
command moves lines, and so on.
It is possible to modify only a portion of a line by means of replacement,
as in the example above.  However even here, the
.I `s'
command is applied to whole lines at a time.

In general,
.B ed
commands consist of zero or more line addresses, followed by a single
character command and possibly additional parameters; i.e.,
commands have the structure:
.sp
.RS
.I [address [,address]]command[parameters]
.RE
.sp
The address(es) indicate the line or range of lines to be affected by the
command.  If fewer addresses are given than the command accepts, then
default addresses are supplied.

.SS OPTIONS
.TP 8
-G
Forces backwards compatibility.  Affects the commands
.IR `G' ,
.IR `V' ,
.IR `f' ,
.IR `l' ,
.IR `m' ,
.IR `t' ,
and
.IR `!!' .
.TP 8
-s
Suppresses diagnostics. This should be used if
.BR ed 's
standard input is from a script.

.TP 8
.RI \-p \ string
Specifies a command prompt.  This may be toggled on and off with the
.I `P'
command.

.TP 8
.I file
Specifies the name of a file to read.  If
.I file
is prefixed with a
bang (!), then it is interpreted as a shell command.  In this case,
what is read is
the standard output of
.I file
executed via
.IR sh (1).
To read a file whose name begins with a bang, prefix the
name with a backslash (\e).
The default filename is set to
.I file
only if it is not prefixed with a bang.

.SS LINE ADDRESSING
An address represents the number of a line in the buffer.
.B ed
maintains a
.I current address
which is
typically supplied to commands as the default address when none is specified.
When a file is first read,  the current address is set to the last line
of the file.  In general, the current address is set to the last line
affected by a command.

A line address is
constructed from one of the bases in the list below, optionally followed
by a numeric offset.  The offset may include any combination
of digits, operators (i.e.,
.IR + ,
.I -
and
.IR ^ )
and whitespace.
Addresses are read from left to right, and their values are computed
relative to the current address.

One exception to the rule that addresses represent line numbers is the
address
.I 0
(zero).
This means "before the first line,"
and is legal wherever it makes sense.

An address range is two addresses separated either by a comma or
semicolon. The value of the first address in a range cannot exceed the
value of the the second.  If only one address is given in a range, then
the second address is set to the given address.  If an
.IR n- tuple
of addresses is given where
.I n > 2,
then the corresponding range is determined by the last two addresses in
the
.IR n- tuple.
If only one address is expected, then the last address is used.

Each address in a comma-delimited range is interpreted relative to the
current address.  In a semicolon-delimited range, the first address is
used to set the current address, and the second address is interpreted
relative to the first.


The following address symbols are recognized.

.TP 8
\&.
The current line (address) in the buffer.

.TP 8
$
The last line in the buffer.

.TP 8
n
The
.IR n th,
line in the buffer
where
.I n
is a number in the range
.I [0,$].

.HP
-
.TP 8
^
The previous line.
This is equivalent to
.I -1
and may be repeated with cumulative effect.

.HP
-\fIn\fR
.TP 8
^\fIn\fR
The
.IR n th
previous line, where
.I n
is a non-negative number.

.TP 8
+
The
next line.
This is equivalent to
.I +1
and may be repeated with cumulative effect.

.HP
+\fIn\fR
.TP 8
whitespace \fIn\fR
The
.IR n th
next line, where
.I n
is a non-negative number.
.I Whitespace
followed by a number
.I n
is interpreted as
.IR +n .

.HP
,
.TP 8
%
The first through last lines in the buffer.  This is equivalent to
the address range
.I 1,$.

.TP 8
;
The current through last lines in the buffer.  This is equivalent to
the address range
.I .,$.

.TP 8
.RI / re/
The
next line containing the regular expression
.IR re .
The search wraps to the beginning of the buffer and continues down to the
current line, if necessary.
// repeats the last search.

.TP 8
.RI ? re?
The
previous line containing the regular expression
.IR re .
The search wraps to the end of the buffer and continues up to the
current line, if necessary.
?? repeats the last search.

.TP 8
.RI \' lc
The
line previously marked by a
.I `k'
(mark) command, where
.I lc
is a lower case letter.

.SS REGULAR EXPRESSIONS
Regular expressions are patterns used in selecting text.
For example, the
.B ed
command
.sp
.RS
g/\fIstring\fR/
.RE
.sp
prints all lines containing
.IR string .
Regular expressions are also
used by the
.I `s'
command for selecting old text to be replaced with new.

In addition to a specifying string literals, regular expressions can
represent
classes of strings.  Strings thus represented are said to be matched
by the corresponding regular expression.
If it is possible for a regular expression
to match several strings in a line, then the left-most longest match is
the one selected.

The following symbols are used in constructing regular expressions:

.TP 8
c
Any character
.I c
not listed below, including `{', '}', `(', `)', `<' and `>',
matches itself.

.TP 8
\e\fIc\fR
A backslash-escaped character
.IR c
other than `{', '}', `(', `)', `<', `>',
`b', 'B', `w', `W', `+', and `?'
matches itself.

.TP 8
\fR.\fR
Matches any single character.

.TP 8
.I [char-class]
Matches any single character in
.IR char-class .
To include a  `]'
in
.IR char-class ,
it must be the first character.
A range of characters may be specified by separating the end characters
of the range with a `-', e.g., `a-z' specifies the lower case characters.
The following literal expressions can also be used in
.I char-class
to specify sets of characters:
.sp
\ \ [:alnum:]\ \ [:cntrl:]\ \ [:lower:]\ \ [:space:]
.PD 0
\ \ [:alpha:]\ \ [:digit:]\ \ [:print:]\ \ [:upper:]
.PD 0
\ \ [:blank:]\ \ [:graph:]\ \ [:punct:]\ \ [:xdigit:]
.sp
If `-' appears as the first or last
character of
.IR char-class ,
then it matches itself.
All other characters in
.I char-class
match themselves.
.sp
Patterns in
.I char-class
of the form:
.sp
\ \ [.\fIcol-elm\fR.] or,
.PD 0
\ \ [=\fIcol-elm\fR=]
.sp
where
.I col-elm
is a
.I collating element
are interpreted according to
.IR locale (5)
(not currently supported).
See
.IR regex (3)
for an explanation of these constructs.

.TP 8
[^\fIchar-class\fR]
Matches any single character, other than newline, not in
.IR char-class .
.IR char-class
is defined
as above.

.TP 8
^
If `^' is the first character of a regular expression, then it
anchors the regular expression to the beginning of a line.
Otherwise, it matches itself.

.TP 8
$
If `$' is the last character of a regular expression, it
anchors the regular expression to the end of a line.
Otherwise, it matches itself.

.TP 8
\e(\fIre\fR\e)
Defines a (possibly null) subexpression
.IR re .
Subexpressions may be nested.
A subsequent backreference of the form `\e\fIn\fR', where
.I n

contrib/gfx/mesa/lib/widgets-mesa/demos/ed.c
/* ed.c -- Event Display (demo program for the MesaWS widget)
   Copyright (C) 1995 Thorsten.Ohl @ Physik.TH-Darmstadt.de

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2 of the License, or
   (at your option) any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program; see the file COPYING.  If not, write to
   the Free Software Foundation, 675 Mass Ave, Cambridge, MA 02139, USA.

   $Id$
 */

/* This is a simple event display as a demonstration of the MesaWS widget.
   It reads a stream of simple event records from standard input (see the
   file `events' for a example).

   It is a slow and crude hack.  */
  
#include <stdio.h>
#include <stdlib.h>
#include <stdarg.h>
#include <math.h>
#ifndef M_PI
#define M_PI 3.14159265358979323846
#endif
#include <X11/X.h>
#include <X11/Intrinsic.h>
#include <X11/StringDefs.h>
#include <X11/Shell.h>
#ifndef __GLX_MOTIF
#include <X11/Xaw/Box.h>
#include <X11/Xaw/Command.h>
#include <X11/Xaw/Form.h>
#include <GL/xmesa.h>
#include <GL/gl.h>
#include <GL/glu.h>
#include <GL/glx.h>
#include <GL/MesaDrawingArea.h>
#include <GL/MesaWorkstation.h>
#else /* __GLX_MOTIF */
#include <Xm/RowColumn.h>
#include <Xm/Label.h>
#include <Xm/PushB.h>
#include <Xm/Form.h>
#include <GL/xmesa.h>
#include <GL/gl.h>
#include <GL/glu.h>
#include <GL/glx.h>
#include <GL/MesaMDrawingArea.h>
#include <GL/MesaMWorkstation.h>
#define GLwMakeCurrent GLwMMakeCurrent
#define GLwPostObject		   GLwMPostObject
#define GLwBeginView		   GLwMBeginView
#define GLwEndView		   GLwMEndView
#define GLwSetPolarView		   GLwMSetPolarView
#define GLwBeginProjection	   GLwMBeginProjection
#define GLwEndProjection	   GLwMEndProjection
#define GLwSetFrustumProjection	   GLwMSetFrustumProjection
#define GLwGetProjectionList	   GLwMGetProjectionList
#define GLwRedrawObjects	   GLwMRedrawObjects
#define GLwPostProjection	   GLwMPostProjection
#endif /* __GLX_MOTIF */

static char *RCS_Id =
"@(#) $Id$";

static GLint Black, White, Red, Green, Blue, Yellow;
Widget mesa, mesax, mesay, mesaz;

static void quit_function (Widget, XtPointer, XtPointer);
static void next_function (Widget, XtPointer, XtPointer);
static void next_event (void);
static void update_angles (float, float);
static void update_number (int);

static void
Quit (Widget w, XEvent *event, String *argv, Cardinal *argc)
{
  quit_function (NULL, NULL, NULL);
}

void
Next (Widget w, XEvent *event, String *argv, Cardinal *argc)
{
  next_function (NULL, NULL, NULL);
}

static XtActionsRec actions[] =
{
  {"Quit", Quit },
  {"Next", Next },
};

void
quit_function (Widget w, XtPointer closure, XtPointer call_data)
{
  exit (0);
}

void
next_function (Widget w, XtPointer closure, XtPointer call_data)
{
  next_event ();
  GLwRedrawObjects (mesa);
  GLwRedrawObjects (mesax);
  GLwRedrawObjects (mesay);
  GLwRedrawObjects (mesaz);
}

#define SET_POLAR(_p, _r, _theta, _phi) \
    (_p)[0] = (_r) * sin (_theta) * cos (_phi), \
    (_p)[1] = (_r) * sin (_theta) * sin (_phi), \
    (_p)[2] = (_r) * cos (_theta)
static void
draw_barrel_segment (float bottom_radius, float top_radius,
		     float front_angle, float back_angle,
		     float half_angle)
{
  int i;
  GLdouble top[4][3], bottom[4][3], *front[4], *back[4];

  SET_POLAR (top[0], top_radius, front_angle, half_angle);
  SET_POLAR (top[1], top_radius, front_angle, -half_angle);
  SET_POLAR (top[2], top_radius, back_angle, -half_angle);
  SET_POLAR (top[3], top_radius, back_angle, half_angle);

  SET_POLAR (bottom[0], bottom_radius, front_angle, half_angle);
  SET_POLAR (bottom[1], bottom_radius, front_angle, -half_angle);
  SET_POLAR (bottom[2], bottom_radius, back_angle, -half_angle);
  SET_POLAR (bottom[3], bottom_radius, back_angle, half_angle);

  front[0] = top[0];
  front[1] = top[1];
  front[2] = bottom[1];
  front[3] = bottom[0];

  back[0] = top[3];
  back[1] = top[2];
  back[2] = bottom[2];
  back[3] = bottom[3];

  glBegin (GL_LINE_LOOP);
    for (i = 0; i <= 3; i++)
      glVertex3dv (top[i]);
  glEnd ();
  
  glBegin (GL_LINE_LOOP);
    for (i = 0; i <= 3; i++)
      glVertex3dv (bottom[i]);
  glEnd ();
  
  glBegin (GL_LINE_LOOP);
    for (i = 0; i <= 3; i++)
      glVertex3dv (front[i]);
  glEnd ();
  
  glBegin (GL_LINE_LOOP);
    for (i = 0; i <= 3; i++)
      glVertex3dv (back[i]);
  glEnd ();
}

static GLuint particles;
static GLuint greek, latin, desc, desc_number, desc_angles;

static void
next_event (void)
{
  char buffer[100];
  enum
    {
      SKIP, READ, DONE
    }
  state;
  int event = 0;
  
  glNewList (particles, GL_COMPILE);
  glBegin (GL_LINES);
  {
    state = SKIP;
    while (state != DONE)
      {
	int id;
	float px, py, pz;
	if (fgets (buffer, 100, stdin) == NULL)
	  state = DONE;
	switch (state)
	  {
	  case SKIP:
	    if (sscanf (buffer, "BEGIN %d", &event) == 1)
	      state = READ;
	    break;
	  case READ:
	    if (sscanf (buffer, "%d %g %g %g",
			&id, &px, &py, &pz) == 4)
	      {
		if (id == 22)
		  {
		    /* Photons */
		    glIndexi (Red);
		    glColor3f (1.0, 0.0, 0.0);
		  }
		else if ((11 == abs (id))
			 || (13 == abs (id))
			 || (15 == abs (id)))
		  {
		    /* Charged leptons */
		    glIndexi (Green);
		    glColor3f (0.0, 1.0, 0.0);
		  }
		else if ((12 == abs (id))
			 || (14 == abs (id))
			 || (16 == abs (id)))
		  {
		    /* Neutral leptons */
		    glIndexi (Yellow);
		    glColor3f (1.0, 1.0, 0.0);
		  }
		else if ((abs (id) <= 6) || (100 <= abs (id)))
		  {
		    /* Quarks/Hadrons */
		    glIndexi (Blue);
		    glColor3f (0.0, 0.0, 1.0);
		  }
		else
		  {
		    /* ??? */
		    glIndexi (White);
		    glColor3f (1.0, 1.0, 1.0);
		  }
		glVertex3f (0.0, 0.0, 0.0);
		glVertex3f (px, py, pz);
	      }
	    else if (strncmp ("END", buffer, 3) == 0)
	      state = DONE;
	    break;
	  case DONE:
	    /* never happens */
	    break;
	  }
      }
  }
  glEnd ();
  glEndList ();

  update_number (event);
}

static void
update_angles (float theta, float phi)
{
  char msg[64];
  sprintf (msg, "f=%5.2fp, q=%5.2fp", phi/M_PI, theta/M_PI);
  glNewList (desc_angles, GL_COMPILE);
    glCallLists (strlen (msg), GL_BYTE, msg);
  glEndList ();
}  

static void
update_number (int event)
{
  char msg[32];
  sprintf (msg, "ADLO/TH, #%d", event);
  glNewList (desc_number, GL_COMPILE);
    glCallLists (strlen (msg), GL_BYTE, msg);
  glEndList ();
}  

static void
setup_context (Widget w)
{
  GLwMakeCurrent (w);
  glClearColor (0.0, 0.0, 0.0, 0.0);
  glClearIndex (Black);
  glShadeModel (GL_FLAT);
}

static GLint
alloc_color (Widget w, Colormap cmap, int red, int green, int blue)
{ 
  XColor xcolor;
  xcolor.red = red;
  xcolor.green = green;
  xcolor.blue = blue;
  xcolor.flags = DoRed | DoGreen | DoBlue;
  if (!XAllocColor (XtDisplay (w), cmap, &xcolor))
    {
      printf ("Couldn't allocate color!\n");
      exit (1);
    }
  return xcolor.pixel;
}

/* This is rather inefficient, but we don't mind for the moment,
   because it works.  */

static void
translate_pixels (Widget to, Widget from, ...)
{
  va_list ap;
  char *name;
  Colormap from_cmap, to_cmap;
  XColor xcolor;

  XtVaGetValues (from, XtNcolormap, &from_cmap, NULL);
  XtVaGetValues (to, XtNcolormap, &to_cmap, NULL);

  va_start (ap, from);
  for (name = va_arg (ap, char *); name != NULL; name = va_arg (ap, char *))
    {
      XtVaGetValues (from, name, &xcolor.pixel, NULL);
      XQueryColor (XtDisplay (from), from_cmap, &xcolor);
      if (!XAllocColor (XtDisplay (to), to_cmap, &xcolor))
	XtAppWarning (XtWidgetToApplicationContext (to),
		      "Couldn't allocate color!\n");
      else
	XtVaSetValues (from, name, xcolor.pixel, NULL);
    }
  va_end (ap);
}

static Widget
create_label (Widget parent, char *name, char *text)
{
#ifndef __GLX_MOTIF
  return XtVaCreateManagedWidget (name, labelWidgetClass, parent,
				  XtNlabel, text, NULL);
#else /* __GLX_MOTIF */
  XmString label = XmStringCreateLtoR (text, XmFONTLIST_DEFAULT_TAG);
  return XtVaCreateManagedWidget (name, xmLabelWidgetClass, parent,
				  XmNlabelString, label, NULL);
  XmStringFree (label);
#endif /* __GLX_MOTIF */
}

static Widget
create_ok (Widget parent, char *name, XtCallbackProc cb)
{
  Widget ok;
#ifndef __GLX_MOTIF
  ok = XtVaCreateManagedWidget (name, commandWidgetClass, parent,
				XtNlabel, "OK", NULL);
  XtAddCallback (ok, XtNcallback, cb, NULL);
#else /* __GLX_MOTIF */
  XmString label = XmStringCreate ("OK", XmFONTLIST_DEFAULT_TAG);
  ok = XtVaCreateManagedWidget (name, xmPushButtonWidgetClass, parent,
				XmNlabelString, label, NULL);
  XmStringFree (label);
  XtAddCallback (ok, XmNarmCallback, cb, NULL);
#endif /* __GLX_MOTIF */
  return ok;
}

static Widget
create_command (Widget parent, char *name, XtCallbackProc cb)
{
  Widget ok;
#ifndef __GLX_MOTIF
  ok = XtVaCreateManagedWidget (name, commandWidgetClass, parent, NULL);
  XtAddCallback (ok, XtNcallback, cb, NULL);
#else /* __GLX_MOTIF */
  ok = XtVaCreateManagedWidget (name, xmPushButtonWidgetClass, parent, NULL);
  XtAddCallback (ok, XmNarmCallback, cb, NULL);
#endif /* __GLX_MOTIF */
  return ok;
}

static Widget
create_popup (Widget parent, char *name)
{
#ifndef __GLX_MOTIF
  return XtVaCreatePopupShell (name, transientShellWidgetClass, parent, NULL);
#else /* __GLX_MOTIF */
  /* ??? */
  return XtVaCreatePopupShell (name, transientShellWidgetClass, parent, NULL);
#endif /* __GLX_MOTIF */
}

static Widget
create_box (Widget parent, char *name)
{
#ifndef __GLX_MOTIF
  return XtVaCreateManagedWidget (name, boxWidgetClass, parent,
				  XtNorientation, XtorientVertical, NULL);
#else /* __GLX_MOTIF */
  return XtVaCreateManagedWidget (name, xmRowColumnWidgetClass, parent,
				  XmNorientation, XmVERTICAL, NULL);
#endif /* __GLX_MOTIF */
}

   
Widget about_shell, help_shell;

char about_message[] = "\
This is ed, a simple particle physics event display\n\
application and demo for the MesaWorkstation widget.\n\
\n\
Copyright (C) 1995 Thorsten.Ohl @ Physik.TH-Darmstadt.de\n\
\n\
This is free software; see the source for copying conditions.\n\
There is NO warranty; not even for MERCHANTABILITY\n\
or FITNESS FOR A PARTICULAR PURPOSE.";

void
popup_about (Widget w, XtPointer closure, XtPointer call_data)
{
  XtPopup (about_shell, XtGrabNonexclusive);
}

void
popdown_about (Widget w, XtPointer closure, XtPointer call_data)
{
  XtPopdown (about_shell);
}


char help_message[] = "\
Sorry, no online help available yet.\n\
\n\
See the manpage or the source of the MesaWorkstation widget\n\
for the available keystroke translations and their actions.\n\
For a start, use the cursor keys to rotate the object.\n\
\n\
The color code of the particles is as follows:\n\
\n\
  red:    photons\n\
  green:  charged leptons (electrons, muons, taus)\n\
  yellow: neutrinos\n\
  blue:   hadrons (incl. quarks)\n\
  while:  unidentified\n";

void
popup_help (Widget w, XtPointer closure, XtPointer call_data)
{
  XtPopup (help_shell, XtGrabNonexclusive);
}

void
popdown_help (Widget w, XtPointer closure, XtPointer call_data)
{
  XtPopdown (help_shell);
}


static XrmOptionDescRec options[] =
{
  { "-debug", "*debug", XrmoptionNoArg, "True" },
  { "-rgba", "*rgba", XrmoptionNoArg, "True" },
  { "-doublebuffer", "*doublebuffer", XrmoptionNoArg, "True" },
  { "-ximage", "*ximage", XrmoptionNoArg, "True" },
};

static String fallback_resources[] =
{
#ifndef __GLX_MOTIF
  "*MesaWorkstation.debug: false",
  "*MesaWorkstation.rgba: false",
  "*MesaWorkstation.installColormap: true",
  "*MesaWorkstation.doublebuffer: true",
  "*MesaWorkstation.ximage: false",
  "*MesaWorkstation.translations: #augment \
   <Key>q: Quit()\n\
   <Key>Return: Next()\n\
   <Key>space: Next()",
#else /* __GLX_MOTIF */
  "*MesaMWorkstation.debug: false",
  "*MesaMWorkstation.rgba: false",
  "*MesaMWorkstation.installColormap: true",
  "*MesaMWorkstation.doublebuffer: true",
  "*MesaMWorkstation.ximage: false",
  "*MesaMWorkstation.translations: #augment \
   <Key>q: Quit()\n\
   <Key>Return: Next()\n\
   <Key>space: Next()",
#endif /* __GLX_MOTIF */
  "*help.label: Help",
  "*about.label: About ed",
  "*quit.label: Exit",
  "*next.label: Next event",
  "*mesa.width: 400",
  "*mesa.height: 400",
  "*mesax.width: 100",
  "*mesax.height: 100",
  "*mesay.width: 100",
  "*mesay.height: 100",
  "*mesaz.width: 100",
  "*mesaz.height: 100",
  NULL
};

int
main (int argc, char *argv[])
{
  Widget top, frame, commands, next, quit;
  Widget about, about_box, about_text, about_id, about_ok;
  Widget help, help_box, help_text, help_ok;
  XtAppContext app_context;
  Boolean rgba, cmap_installed;
  
  GLuint barrel, object;

  XtSetLanguageProc (NULL, NULL, NULL);
  top = XtVaAppInitialize (&app_context, "Ed",
			   options, XtNumber (options),
			   &argc, argv, fallback_resources,
			   NULL);

#ifndef __GLX_MOTIF
  frame = XtVaCreateManagedWidget ("frame", formWidgetClass,
				   top,
				   NULL);
  commands = XtVaCreateManagedWidget ("commands", boxWidgetClass,
				      frame,
				      XtNhSpace, 10,
				      XtNorientation, XtorientHorizontal,
				      NULL);
#else /* __GLX_MOTIF */
  frame = XtVaCreateManagedWidget ("frame", xmFormWidgetClass,
				   top,
				   /* This should not be necessary.  Is it
				      a bug in LessTif or in my understanding
				      of Motif?  */
				   XtNwidth, 420, XtNheight, 600,
 				   NULL);
  commands = XtVaCreateManagedWidget ("commands", xmRowColumnWidgetClass,
				      frame,
				      XmNorientation, XmHORIZONTAL,
				      XmNtopAttachment, XmATTACH_FORM,
				      XmNtopOffset, 10,
				      XmNrightAttachment, XmATTACH_FORM,
				      XmNrightOffset, 10,
				      NULL);
#endif /* __GLX_MOTIF */

  help = create_command (commands, "help", popup_help);
  help_shell = create_popup (top, "help_shell");
  help_box = create_box (help_shell, "help_box");
  help_text = create_label (help_box, "help_text", help_message);
  help_ok = create_ok (help_box, "help_ok", popdown_help);

  about = create_command (commands, "about", popup_about);
  about_shell = create_popup (top, "about_shell");
  about_box = create_box (about_shell, "about_box");
  about_text = create_label (about_box, "about_text", about_message);
  about_id = create_label (about_box, "about_id", RCS_Id);
  about_ok = create_ok (about_box, "about_ok", popdown_about);

  next = create_command (commands, "next", next_function);
  quit = create_command (commands, "quit", quit_function);

#ifndef __GLX_MOTIF
  mesa = XtVaCreateManagedWidget ("mesa", mesaWorkstationWidgetClass,
				  frame,
				  GLwNshareLists, True,
				  XtNfromVert, commands, XtNvertDistance, 10,
				  NULL);
  mesax = XtVaCreateManagedWidget ("mesax", mesaWorkstationWidgetClass,
				   frame,
				   GLwNshareLists, True,
				   XtNfromVert, mesa, XtNvertDistance, 10,
				   NULL);
  mesay = XtVaCreateManagedWidget ("mesay", mesaWorkstationWidgetClass,
				   frame,
				   GLwNshareLists, True,
				   XtNfromVert, mesa, XtNvertDistance, 10,
				   XtNfromHoriz, mesax, XtNhorizDistance, 50,
				   NULL);
  mesaz = XtVaCreateManagedWidget ("mesaz", mesaWorkstationWidgetClass,
				   frame,
				   GLwNshareLists, True,
				   XtNfromVert, mesa, XtNvertDistance, 10,
				   XtNfromHoriz, mesay, XtNhorizDistance, 50,
				   NULL);
#else /* __GLX_MOTIF */
  mesa = XtVaCreateManagedWidget ("mesa", mesaMWorkstationWidgetClass,
				  frame,
				  GLwNshareLists, True,
				  XmNtopAttachment, XmATTACH_WIDGET,
				  XmNtopOffset, 10,
				  XmNtopWidget, commands,
				  XmNleftAttachment, XmATTACH_FORM,
				  XmNleftOffset, 10,
				  XmNrightAttachment, XmATTACH_FORM,
				  XmNrightOffset, 10,
				  NULL);
  mesax = XtVaCreateManagedWidget ("mesax", mesaMWorkstationWidgetClass,
				   frame,
				   GLwNshareLists, True,
				   XmNtopAttachment, XmATTACH_WIDGET,
				   XmNtopOffset, 10,
				   XmNtopWidget, mesa,
				   XmNleftAttachment, XmATTACH_FORM,
				   XmNleftOffset, 10,
				   XmNbottomAttachment, XmATTACH_FORM,
				   XmNbottomOffset, 10,
				   NULL);
  mesay = XtVaCreateManagedWidget ("mesay", mesaMWorkstationWidgetClass,
				   frame,
				   GLwNshareLists, True,
				   XmNtopAttachment, XmATTACH_WIDGET,
				   XmNtopOffset, 10,
				   XmNtopWidget, mesa,
				   XmNleftAttachment, XmATTACH_WIDGET,
				   XmNleftOffset, 50,
				   XmNleftWidget, mesax,
				   XmNbottomAttachment, XmATTACH_FORM,
				   XmNbottomOffset, 10,
				   NULL);
  mesaz = XtVaCreateManagedWidget ("mesaz", mesaMWorkstationWidgetClass,
				   frame,
				   GLwNshareLists, True,
				   XmNtopAttachment, XmATTACH_WIDGET,
				   XmNtopOffset, 10,
				   XmNtopWidget, mesa,
				   XmNleftAttachment, XmATTACH_WIDGET,
				   XmNleftOffset, 50,
				   XmNleftWidget, mesay,
				   XmNbottomAttachment, XmATTACH_FORM,
				   XmNbottomOffset, 10,
				   XmNrightAttachment, XmATTACH_FORM,
				   XmNrightOffset, 10,
				   NULL);
#endif /* __GLX_MOTIF */

  XtAppAddActions (app_context, actions, XtNumber (actions));
  
  XtRealizeWidget (top);

  XtVaGetValues (mesa,
		 GLwNrgba, &rgba,
		 GLwNinstallColormap, &cmap_installed,
		 NULL);
  if (rgba)
    {
      Black = White = Red = Green = Blue = Yellow = 0;

      if (cmap_installed)
	{
	  /* In RGBA mode, the Mesa widgets will have their own color map.
	     Adjust the colors of the other widgets so that--even if the rest
	     of the screen has wrong colors--all application widgets have the
	     right colors.  */
		     
	  translate_pixels (mesa, quit,
			    XtNbackground, XtNforeground, XtNborder, NULL);
	  translate_pixels (mesa, next,
			    XtNbackground, XtNforeground, XtNborder, NULL);
	  translate_pixels (mesa, help,
			    XtNbackground, XtNforeground, XtNborder, NULL);
	  translate_pixels (mesa, about,
			    XtNbackground, XtNforeground, XtNborder, NULL);
	  translate_pixels (mesa, commands, XtNbackground, XtNborder, NULL);
	  translate_pixels (mesa, frame, XtNbackground, XtNborder, NULL);
	  
	  /* Finally warp the pointer into the mesa widget, to make sure that
	     the user sees the right colors at the beginning.  */

	  XWarpPointer (XtDisplay (mesa), None, XtWindow (mesa), 0, 0, 0, 0, 0, 0);
	}
    }
  else
    {
      /* Allocate a few colors for use in color index mode.  */

      Colormap cmap;
      cmap = DefaultColormap (XtDisplay (top), DefaultScreen (XtDisplay (top)));
      Black  = alloc_color (top, cmap, 0x0000, 0x0000, 0x0000);
      White  = alloc_color (top, cmap, 0xffff, 0xffff, 0xffff);
      Red    = alloc_color (top, cmap, 0xffff, 0x0000, 0x0000);
      Green  = alloc_color (top, cmap, 0x0000, 0xffff, 0x0000);
      Blue   = alloc_color (top, cmap, 0x0000, 0x0000, 0xffff);
      Yellow = alloc_color (top, cmap, 0xffff, 0xffff, 0x0000);
    }

  setup_context (mesa);
  setup_context (mesax);
  setup_context (mesay);
  setup_context (mesaz);

  GLwSetFrustumProjection (mesa, -1.0, 1.0, -1.0, 1.0, 1.0, 10.0);
  GLwSetPolarView (mesa, 3, M_PI/2, 0.0);

  GLwBeginProjection (mesax);
  glOrtho (-1.5, 1.5, -1.5, 1.5, 1.0, 10.0);
  GLwEndProjection ();
  GLwPostProjection (mesay, GLwGetProjectionList (mesax));
  GLwPostProjection (mesaz, GLwGetProjectionList (mesax));

  GLwBeginView (mesax);
  gluLookAt (3.0, 0.0, 0.0,  0.0, 0.0, 0.0,  0.0, 1.0, 0.0);
  GLwEndView ();

  GLwBeginView (mesay);
  gluLookAt (0.0, 3.0, 0.0,  0.0, 0.0, 0.0,  0.0, 0.0,-1.0);
  GLwEndView ();

  GLwBeginView (mesaz);
  gluLookAt (0.0, 0.0, 3.0,  0.0, 0.0, 0.0,  0.0, 1.0, 0.0);
  GLwEndView ();

  barrel = glGenLists (1);
  glNewList (barrel, GL_COMPILE);
  {
    draw_barrel_segment (0.9, 1.0,
			 0.3 * M_PI, 0.7 * M_PI,
			 0.1 * M_PI);
  }
  glEndList ();

  particles = glGenLists (1);
  glNewList (particles, GL_COMPILE);
  glEndList ();

  object = glGenLists (1);
  glNewList (object, GL_COMPILE);
  {
    float frac;
    glClear (GL_COLOR_BUFFER_BIT);
    glIndexi (White);
    glColor3f (1.0, 1.0, 1.0);
    for (frac = 0; frac < 1.0; frac += 0.125)
      {
	glPushMatrix ();
	  glRotatef (frac * 360, 0.0, 0.0, 1.0);
	  glCallList (barrel);
	glPopMatrix ();
      }
    glPushMatrix ();
      glCallList (particles);
    glPopMatrix ();
  }
  glEndList ();

#if 0
  greek = glGenLists (128);
  glXUseXFont (XLoadFont (XtDisplay (top), "*-symbol-*-240-*"),
	       0, 128, greek);
#endif
  latin = glGenLists (128);
  glXUseXFont (XLoadFont (XtDisplay (top), "*-times-*-240-*"),
	       0, 128, latin);

  desc_number = glGenLists (1);
  update_number (0);
  desc_angles = glGenLists (1);
  update_angles (0.0, 0.0);

  desc = glGenLists (1);
  glNewList (desc, GL_COMPILE);
  {
    glMatrixMode (GL_PROJECTION);
    glLoadIdentity ();
    glOrtho (0, 400, 0, 400, -1.0, 1.0);
    glMatrixMode (GL_MODELVIEW);
    glLoadIdentity ();
    glIndexi (White);
    glColor3f (1.0, 1.0, 1.0);
    glPushAttrib (GL_LIST_BIT);
      glRasterPos2i (10.0, 10.0);
      glListBase (latin);
      glCallList (desc_number);
#if 0
      glRasterPos2i (10.0, 380.0);
      glListBase (greek);
      glCallList (desc_angles);
#endif
    glPopAttrib ();
  }
  glEndList ();
  
  GLwPostObject (mesa, object);
  GLwPostObject (mesax, object);
  GLwPostObject (mesay, object);
  GLwPostObject (mesaz, object);
  GLwPostObject (mesa, desc);

  XtAppMainLoop (app_context);
  return (0);
}

